### PR TITLE
Route through the updater for unv:// protocol handling

### PR DESCRIPTION
**Pull request by slipher**
The updater now takes a positional command line argument. This is the URL of an Unvanquished server to connect to. The updater is now used for the protocol handler on supported platforms. This allows one of the Linux desktop files to be removed. When the next version of Daemon is released, the updater will start to respect custom command lines when there is a URL.

Also fix a race condition with the Linux desktop database.

TODO: test the case of plumbing a URL through an updater update

### CODE REVIEW

File: resources/net.unvanquished.Unvanquished.desktop [hunk 1]
```diff
@@ -5,8 +5,9 @@ Comment=FPS/RTS Game - Aliens vs. Humans
 Icon=unvanquished
 Terminal=false
 Type=Application
-Exec="%1/updater"
+Exec="%1/updater" %u
```
Comment (DolceTriade) [inline resources/net.unvanquished.Unvanquished.desktop]: Do you think we should do

`updater <updater args> -- <unvanquished args>`

This way, we can always disambiguate updater args from Unvanquished args?

**Review by DolceTriade — APPROVED** [review]
Otherwise, LGTM

File: resources/net.unvanquished.Unvanquished.desktop (hunk 1 above)
Comment (illwieckz) [reply resources/net.unvanquished.Unvanquished.desktop]: This is a very good idea.

Comment (slipher) [reply resources/net.unvanquished.Unvanquished.desktop]: Done.

File: main.cpp [hunk 2]
```diff
@@ -60,8 +62,55 @@ struct CommandLineOptions {
...
+}
+
+QString getURLFromPositionalOptions(const QCommandLineParser& parser)
+{
+    QStringList args = parser.positionalArguments();
+    if (args.empty()) {
+        return "";
+    } else if (args.size() > 1) {
+        argParseError("Too many command line arguments.");
+    }
```
Comment (DolceTriade) [inline main.cpp:81]: Is this correct with the --?

File: main.cpp [hunk 3]
```diff
@@ -60,8 +62,55 @@ struct CommandLineOptions {
...
+    // leading '['.
+    if (url[passwordEnd] == '[') {
+        url.remove(passwordEnd, 1);
+    }
+
+    // If a custom command is used, we have to consider the pessimistic case
+    // of a command line like   bash -c "%command%"
+    // so metacharacters from any common shells must be forbidden.
+    // What we need to accept are domain names and IP addresses.
+    // Also throw in '@' for limited support of the password feature (unv://hunter2@1.2.3.4)
```
Comment (DolceTriade) [inline main.cpp:103]: Should we not just have an option to shell escape stuff? This seems brittle and hacky.

File: main.cpp (hunk 2 above)
Comment (slipher) [reply main.cpp:81]: I tested it.

File: main.cpp (hunk 3 above)
Comment (slipher) [reply main.cpp:103]: Most of the time it won't be in a shell context, so if it is escaped you won't get the right result.

Another approach could be implementing URL `%` decoding in Daemon for "special" characters. This seems not that different from a shell safety perspective (no longer allowing `]@` but allowing `%`). The advantages of that would be avoiding the `[` erasing hack and allowing any character in the password. Disadvantages would be not working for the current Daemon release and needing to do percent decoding in the updater too for it to forward correctly.

Comment (slipher) [reply main.cpp:103]: Actually that's probably worse since `%` is a shell metacharacter in Windows.

File: main.cpp (hunk 2 above)
Comment (DolceTriade) [reply main.cpp:81]: Why does that work? The QT thing make `--` magical?

Comment (slipher) [reply main.cpp:81]: Yes, the Qt arg parser enables `--` handling by default.

**Comment by slipher** [comment]
Added a new commit to make the URL handler feature still work if there is an updater update.

File: main.cpp (hunk 2 above)
Comment (DolceTriade) [reply main.cpp:81]: Ah interesting.

**Review by DolceTriade — APPROVED** [review]

**Pull request merged**